### PR TITLE
GP functionality 2

### PR DESCRIPTION
**Pull request by bwengals**
I refactored the design of the GP module.  To keep the PR clean, I'm closing #2322 in favor of this.  See a description of what I'm trying to do [here](https://bwengals.github.io/big-refactor.html).  Feeling much better about how this is set up now.  If anyone has feedback or suggestions on API / how things are named, I'd love to hear.   ping @fonnesbeck, @AustinRochford 

Also, I took `cov.py` and `mean.py` directly from #2322, so you've probably seen these changes already.

ref: #2428

### CODE REVIEW

**Comment by junpenglao** [comment]
wow @bwengals your blog is really well written, the refactoring looks awesome! 👍

File: pymc3/gp/gp.py [hunk 1]
```diff
@@ -1,151 +1,308 @@
...
+        self.size = size
+        self.mean_func = mean_func
 
-        if mean_func is None:
-            self.M = Zero()
+        # force user to run `__call__` before `conditioned_on`
+        self._ready = True
+        return self
+
+    def conditioned_on(self, X, Xs=None, **kwargs):
```
Comment (ferrine) [inline pymc3/gp/gp.py]: Very strange choice of API. Do you have a brief example of usage?

File: pymc3/gp/gp.py [hunk 2]
```diff
@@ -1,151 +1,308 @@
...
+        if self._ready:
+            self._ready = False
```
Comment (ferrine) [inline pymc3/gp/gp.py]: It is fragile to possible bugs below. Will it be ok if code below breaks at some point and `self._ready = False`?

**Review by ferrine — COMMENTED** [review]
I really liked your generic `GP` function. It seems to me you use another approach here. I wanted to suggest you to use factory `GP` class that chooses the right child class in `__new__` method. You can see some reference to this architecture choice [here](http://myownhat.blogspot.ru/2012/03/polymorphism-beyond-factory-method.html)

File: pymc3/gp/gp.py (hunk 1 above)
Comment (junpenglao) [reply pymc3/gp/gp.py]: There is some example in the blog https://bwengals.github.io/big-refactor.html

File: pymc3/gp/gp.py (hunk 2 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: Hmm that's true, the `_ready` check is brittle to that.  It's mostly there so that an informative error msg is raised when `__call__` isn't run before `conditioned_on`.  If it weren't there, I think the user would end up getting runtime errors when trying to access stuff like self.size, or when trying to construct a second random variable with the same name.

**Comment by bwengals** [comment]
hey thanks @junpenglao!  

@ferrine The factory approach was great, but it ends up running into problems when trying let gp objects be additive.   With one interface, specifying two GP's separately that get added together later will have a lot of requirements about being given the same arguments.  Something like

    gp1 = pm.gp.GP("gp1", X1, approx="vfe", ...)
    gp2 = pm.gp.GP("gp2", X2, approx=None, ...)
    gp = gp1 + gp2

won't work and should error for a lot of reasons.  Also, for models with Gaussian noise that have the latent function integrated out
  
    gp1 = pm.gp.GP("gp1", X, sigma, observed=y)
    gp2 = pm.gp.GP("gp2", X, sigma, observed=y)
    gp = gp1 + gp2

isn't accurate.  `gp1` and `gp2` weren't observed with IID noise with stdev sigma, their sum `gp` was.  Running into these problems caused me to think about a different API.  That use for `__new__` is a cool idea though.

**Comment by fonnesbeck** [comment]
Yeah, I'm not sold on having `GPLatent` as a user-facing class. Also, being able to instantiate GPs without names will be confusing given the existing API. Could you not leverage the `.dist()` method that we use with other distributions when constructing an additive GP? 

I do like the `conditioned_on(X)` convention, though I'd be in favor of shortening it to `condition(X)`. I still kind of like the original dict syntax for `observed` where you'd be able to pass `observed={'X': X}` instead of using a methods (since `observed` is how we condition in PyMC3). 

Is something along the lines of the following possible?

```python
    cov1 = pm.gp.ExpQuad(1, ℓ1)
    gp1 = pm.GP.dist(cov_func=cov1)

    cov2 = pm.gp.Matern52(1, ℓ2)
    gp2 = pm.GP.dist(cov_func=cov2)

    full_gp = pm.GP('full_gp', gp1 + gp2, mean_func=pm.Zero()).condition(X)
```

or again, perhaps:

     full_gp = pm.GP('full_gp', gp1 + gp2, mean_func=pm.Zero(), observed=dict(X=X))

Anyhow, this is a very thoughtful refactor and I think something good will come of it.

**Comment by bwengals** [comment]
Thanks! this is why i've been quiet lately...

> Yeah, I'm not sold on having GPLatent as a user-facing class. 

Just to clarify, there's a few user facing classes currently in this proposal, `GPMarginal`, `GPLatent`, `TProcess`, and `GPSparseMarginal`.

`observed` can't be a random variable though, right?  This doesn't come up for the others, but `GPLatent` (I had it called `GPNonConjugate` before) has to be conditioned on `f`, or more specifically, what it was reparameterized to, `v`, 
    
    L = cholesky(K)
    v ~ Normal(0, 1, shape)
    f ~ mu + Lv

Of course the `observed` dict would get handled before being passed to a distribution, but people may be confused by the exception to the rule here.  On the other hand, maybe its not so bad to do this and be clear about it.  `GPLatent` is kind of the troublemaker in general...

I didn't think of using `dist`.    First thought is that I like it a lot.  It would make it easy to test out GP priors / covariance functions without dealing with theano,

    samples = pm.GP.dist(cov_func=pm.ExpQuad(1, lengthscale=1)).random(10)
    plt.plot(samples);

I do think there's a case now for using the specific class of the GP to build it up front.  Referring to the 
`gp1 = pm.GPMarginal(cov_func=...)` part.

- Since it's explicit, someone reading (or writing) the code can figure out what's going on faster than mentally parsing the argument combinations.  They'd just need to call `GPThing?` or `help(GPThing)`.  This also cuts down the number of kwargs that would need to be given.
- It causes gp1 + gp2 to error immediately when gp1 and gp2 don't have the same type.  
- It makes it much easier I think for new GP models/implementations to be contributed to the code base.  No changes to the logic of a general constructor need to be made, just the addition of `GPCoolNewOne`.

File: pymc3/gp/cov.py [hunk 3]
```diff
@@ -88,7 +92,8 @@ def __array_wrap__(self, result):
 class Combination(Covariance):
     def __init__(self, factor_list):
         input_dim = np.max([factor.input_dim for factor in
-                            filter(lambda x: isinstance(x, Covariance), factor_list)])
+                            filter(lambda x: isinstance(x, Covariance),
```
Comment (AustinRochford) [inline pymc3/gp/cov.py]: Our BDFL (Python's not PyMC3's) generally frowns upon `filter` and `map` these days, as much as it makes my Haskell-loving heart cry a bit.  More Pythonic may be:

```python
[x for x in factor_list if isinstance(x, Covariance)]
```

Comment (AustinRochford) [reply pymc3/gp/cov.py]: This approach seems like it would let you collapse the nested iterations into a single comprehension as well.

File: pymc3/gp/cov.py [hunk 4]
```diff
@@ -97,29 +102,69 @@ def __init__(self, factor_list):
...
-                factor_list.append(factor(X, Z, diag))
-            elif hasattr(factor, "ndim"):
-                if diag:
-                    factor_list.append(tt.diag(factor))
-                else:
-                    factor_list.append(factor)
-            else:
-                factor_list.append(factor)
+                factor_list.append(factor(X, Xs, diag))
+                continue
```
Comment (AustinRochford) [inline pymc3/gp/cov.py]: I personally find `continue` a bit harder to read than `elif`, others may disagree.

File: pymc3/gp/cov.py [hunk 5]
```diff
@@ -97,29 +102,69 @@ def __init__(self, factor_list):
...
+            factor_list.append(factor)
+
         return factor_list
 
 
 class Add(Combination):
-    def __call__(self, X, Z=None, diag=False):
-        return reduce((lambda x, y: x + y), self.merge_factors(X, Z, diag))
+    def __call__(self, X, Xs=None, diag=False):
+        return reduce((lambda x, y: x + y), self.merge_factors(X, Xs, diag))
```
Comment (AustinRochford) [inline pymc3/gp/cov.py]: Could save some `lambda`s here and in `Prod` using the `operator` module, but it's fine either way.

File: pymc3/gp/cov.py [hunk 6]
```diff
@@ -136,30 +181,45 @@ def __init__(self, input_dim, lengthscales, active_dims=None):
         super(Stationary, self).__init__(input_dim, active_dims)
         self.lengthscales = tt.as_tensor_variable(lengthscales)
 
-    def square_dist(self, X, Z):
+    def square_dist(self, X, Xs):
```
Comment (AustinRochford) [inline pymc3/gp/cov.py]: A docstring or comment showing how your operations are equivalent to the useful definitions would be useful here.

File: pymc3/gp/cov.py (hunk 5 above)
Comment (fonnesbeck) [reply pymc3/gp/cov.py]: Doesn't `sum([x for x in self.merge_factors(X, Z, diag)])` do the trick here?

Comment (fonnesbeck) [reply pymc3/gp/cov.py]: I guess not because we want the `__add__` method. But, `np.sum(self.merge_factors(X, Z, diag))` would work, right?

File: pymc3/gp/cov.py (hunk 3 above)
Comment (bwengals) [reply pymc3/gp/cov.py]: list comprehensions are functional too tho!  Will change, definitely has more pythonicalaceousness

File: pymc3/gp/cov.py (hunk 6 above)
Comment (bwengals) [reply pymc3/gp/cov.py]: ah that would be good

File: pymc3/gp/cov.py (hunk 5 above)
Comment (bwengals) [reply pymc3/gp/cov.py]: I'll take a look again.  I remember it was tricky but don't remember why exactly.

File: docs/source/gp.rst [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+  Machine Learning <http://www.gaussianprocess.org/gpml/>`_ by Rasmussen &
+  Williams, or `this introduction <https://www.ics.uci.edu/~welling/teaching/KernelsICS273B/gpB.pdf>`_ 
+  by D. Mackay.
+
+=============================
+Overview of GP usage In PyMC3
+=============================
+
+PyMC3 is a great environment for working with fully Bayesian Gaussian Process
+models.  The GP functionality of PyMC3 is meant to have a clear syntax and be
```
Comment (AustinRochford) [inline docs/source/gp.rst]: Don't sell your work short, it is clear and highly composable; you can drop "is meant to" ;)

**Comment by twiecki** [comment]
@bwengals looks like you need to rebase.

**Comment by springcoil** [comment]
@bwengals I've pulled your branch and tried replicating the notebook from https://bwengals.github.io/looking-at-the-keeling-curve-with-gps-in-pymc3.html . But I get `mean_func` undefined - have you not included that variable in the notebook. Everything else worked except that part. I'll park this for now until this is more stable.

File: docs/source/gp.rst (hunk 7 above)
Comment (springcoil) [reply docs/source/gp.rst]: Agreed :)

File: pymc3/gp/gp.py [hunk 8]
```diff
@@ -78,7 +57,7 @@ def conditional(self, name, n_points, Xnew, *args, **kwargs):
 
 
 @conditioned_vars(["X", "f"])
-class Latent(GPBase):
+class Latent(Base):
     """ Where the GP f isnt integrated out, and is sampled explicitly
```
Comment (springcoil) [inline pymc3/gp/gp.py]: I think this could be improved with a bit more doc strings. Maybe just with a simple example from somewhere else.

File: pymc3/gp/gp.py [hunk 9]
```diff
@@ -92,40 +71,46 @@ def _build_prior(self, name, n_points, X, reparameterize=True):
...
+        return mu, cov
 
-    def conditional(self, name, n_points, Xnew, X=None, f=None):
-        if X is None: X = self.X
-        if f is None: f = self.f
-        mu, chol = self._build_conditional(Xnew, X, f)
+    def conditional(self, name, n_points, Xnew, gp=None):
+        X, f, cov_total, mean_total = self._get_cond_vals(gp)
+        mu, cov = self._build_conditional(Xnew, X, f, cov_total, mean_total)
+        chol = cholesky(stabilize(cov))
```
Comment (springcoil) [inline pymc3/gp/gp.py]: Neat!

File: pymc3/gp/gp.py [hunk 10]
```diff
@@ -258,18 +241,9 @@ def predict(self, Xnew, point=None, gp=None, pred_noise=False, diag=False):
...
-    #    return pm.MvNormal(name, mu=mu, chol=chol, shape=n_points)
-
-
-
 
 @conditioned_vars(["X", "Xu", "y", "sigma"])
-class MarginalSparse(GPBase):
+class MarginalSparse(Base):
     _available_approx = ["FITC", "VFE", "DTC"]
     """ FITC and VFE sparse approximations
```
Comment (springcoil) [inline pymc3/gp/gp.py]: Would some more docs here help? Or is that elsewhere in the code?

**Review by springcoil — COMMENTED** [review]
Just some general comments - about clarity etc.

File: docs/source/gp.rst (hunk 7 above)
Comment (bwengals) [reply docs/source/gp.rst]: glad you think so!

File: pymc3/gp/gp.py (hunk 8 above)
Comment (bwengals) [reply pymc3/gp/gp.py]: Definitely agree.  been putting those off till the end in case something needs changing

**Comment by springcoil** [comment]
Seems there are some errors still in tests. 

```
_________________________ TestCovSliceDim.test_slice2 __________________________
self = <pymc3.tests.test_gp.TestCovSliceDim object at 0x7f5e0412e690>
    def test_slice2(self):
        X = np.linspace(0, 1, 30).reshape(10, 3)
        with Model() as model:
            cov = gp.cov.ExpQuad(3, [0.1, 0.1], active_dims=[False, True, True])
>       K = theano.function([], cov(X))()
pymc3/tests/test_gp.py:226: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/gp/cov.py:55: in __call__
    return self.full(X, Xs)
pymc3/gp/cov.py:257: in full
    return tt.exp(-0.5 * self.square_dist(X, Xs))
pymc3/gp/cov.py:207: in square_dist
    X = tt.mul(X, 1.0 / self.lengthscales)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/op.py:674: in __call__
    required = thunk()
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/op.py:843: in rval
    fill_storage()
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/cc.py:1698: in __call__
    reraise(exc_type, exc_value, exc_trace)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tp = <type 'exceptions.ValueError'>
value = ValueError('Input dimension mis-match. (input[0].shape[1] = 3, input[1].shape[1] = 2)',)
tb = None
>   ???
E   ValueError: Input dimension mis-match. (input[0].shape[1] = 3, input[1].shape[1] = 2)
<string>:2: ValueError
_________________________ TestCovSliceDim.test_slice3 __________________________
self = <pymc3.tests.test_gp.TestCovSliceDim object at 0x7f5e04098f90>
    def test_slice3(self):
        X = np.linspace(0, 1, 30).reshape(10, 3)
        with Model() as model:
            cov = gp.cov.ExpQuad(3, np.array([0.1, 0.1]), active_dims=[False, True, True])
>       K = theano.function([], cov(X))()
pymc3/tests/test_gp.py:236: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymc3/gp/cov.py:55: in __call__
    return self.full(X, Xs)
pymc3/gp/cov.py:257: in full
    return tt.exp(-0.5 * self.square_dist(X, Xs))
pymc3/gp/cov.py:207: in square_dist
    X = tt.mul(X, 1.0 / self.lengthscales)
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/op.py:674: in __call__
    required = thunk()
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/op.py:843: in rval
    fill_storage()
../../../miniconda2/envs/testenv/lib/python2.7/site-packages/theano/gof/cc.py:1698: in __call__
    reraise(exc_type, exc_value, exc_trace)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tp = <type 'exceptions.ValueError'>
value = ValueError('Input dimension mis-match. (input[0].shape[1] = 3, input[1].shape[1] = 2)',)
tb = None
>   ???
E   ValueError: Input dimension mis-match. (input[0].shape[1] = 3, input[1].shape[1] = 2)
<string>:2: ValueError
_________________________ TestCovSliceDim.test_raises __________________________
self = <pymc3.tests.test_gp.TestCovSliceDim object at 0x7f5e040ce850>
    def test_raises(self):
        lengthscales = 2.0
        with pytest.raises(ValueError):
            gp.cov.ExpQuad(1, lengthscales, [True, False])
>           gp.cov.ExpQuad(2, lengthscales, [True])
E           Failed: DID NOT RAISE <type 'exceptions.ValueError'>
pymc3/tests/test_gp.py:256: Failed
____________________________ TestGP.test_func_args _____________________________
self = <pymc3.tests.test_gp.TestGP object at 0x7f5e9354d2d0>
    def test_func_args(self):
        X = np.linspace(0, 1, 10)[:, None]
        Y = np.random.randn(10, 1)
        with Model() as model:
            # make a Gaussian model
            with pytest.raises(ValueError):
>               random_test = gp.GP('random_test', cov_func=gp.mean.Zero(), observed={'X':X, 'Y':Y})
E               AttributeError: 'module' object has no attribute 'GP'
pymc3/tests/test_gp.py:474: AttributeError
______________________________ TestGP.test_sample ______________________________
self = <pymc3.tests.test_gp.TestGP object at 0x7f5dfe156650>
    def test_sample(self):
        X = np.linspace(0, 1, 10)[:, None]
        Y = np.random.randn(10)
        with Model() as model:
            M = gp.mean.Zero()
            l = Uniform('l', 0, 5)
            K = gp.cov.Matern32(1, l)
            sigma = Uniform('sigma', 0, 10)
            # make a Gaussian model
>           random_test = gp.GP('random_test', mean_func=M, cov_func=K, sigma=sigma, observed={'X':X, 'Y':Y})
E           AttributeError: 'module' object has no attribute 'GP'
pymc3/tests/test_gp.py:488: AttributeError
``` 
Some of these look similar to the error brought up in https://github.com/pymc-devs/pymc3/issues/2505

File: pymc3/gp/grid.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        if n is None:
+            n = (None, None)
+        if len(X) != 2 or len(n) != 2:
+            raise ValueError("2d implemented")
+        n = (infer_shape(X[0], n[0]), infer_shape(X[1], n[1]))
+        logp = lambda Y: self._build_marginal_likelihood(X, Y, sigma, n)
+        self.X = (tt.as_tensor_variable(X[0]), tt.as_tensor_variable(X[1]))
+        self.n = n
+        return f
+
```
Comment (twiecki) [inline pymc3/gp/grid.py]: excess new-lines. Usually there are editor-plugins for this.

**Comment by twiecki** [comment]
This PR is becoming quite large, what would be a good point to try and merge it? You also need to rebase.

**Comment by fonnesbeck** [comment]
A merge from master is probably safer than rebase here.

**Comment by bwengals** [comment]
@twiecki @fonnesbeck went with merge

**Comment by bwengals** [comment]
Thanks @all for the comments throughout, I've tried to address them.  I think it's ready for a review before merging.

**Comment by junpenglao** [comment]
Great work @bwengals!

**Comment by AustinRochford** [comment]
Looks good

**Comment by junpenglao** [comment]
There are still some test fail?

**Comment by fonnesbeck** [comment]
Yes, looks like some problems with argument ordering under Python 2.7.

Why are there more 2.7 tests than 3.6?

**Comment by bwengals** [comment]
oh thats true, tested locally on 3.5 before pushing.

**Comment by ferrine** [comment]
Code is superclean, love your codestyle

**Comment by junpenglao** [comment]
@bwengals In the GP-MaunaLoa.ipynb example, are you using the find_MAP() from the master or your local branch?

**Comment by bwengals** [comment]
Yes it is.  For the life of me i couldnt get it to converge to a result that wasnt junk with the version of findmap on master

**Comment by junpenglao** [comment]
@bwengals LOL, well you might need to PR again the `find_MAP` edit... I pushed another PR in between.

**Comment by bwengals** [comment]
I know.. got a little boxed in there

**Comment by twiecki** [comment]
https://github.com/bwengals/pymc3/blob/7ae8cc3484cf0277577b4f44bfcff3feb2308ccc/docs/source/notebooks/GP-MaunaLoa.ipynb misses the plots

**Comment by bwengals** [comment]
@twiecki maybe thats just githubs notebook reader.  I see the plots in a local build of the docs.  Not sure if they'll come out on whatever is hosting the docs website though.

**Comment by ColCarroll** [comment]
https://stackoverflow.com/questions/39095677/why-doesnt-bokehjs-show-in-a-rendered-jupyter-ipython-notebook-on-github

tl;dr: github scrubs javascript before displaying, so bokeh doesn't get previews.  Should work fine on docs website.

**Comment by bwengals** [comment]
Ah, thanks @ColCarroll!

File: pymc3/gp/gp.py [hunk 12]
```diff
@@ -1,151 +1,753 @@
...
+
+        # A one dimensional column vector of inputs.
+        X = np.linspace(0, 1, 10)[:, None]
+
+        with pm.Model() as model:
+            # Specify the covariance function.
+            cov_func = pm.gp.cov.ExpQuad(1, ls=0.1)
+
+            # Specify the GP.  The default mean function is `Zero`.
+            gp = pm.gp.Latent(cov_func=cov_func)
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py:338]: Shouldn't this be the `Marginal` class?

File: pymc3/gp/gp.py [hunk 13]
```diff
@@ -1,151 +1,753 @@
...
+        A = solve_lower(L, Kxs)
+        v = solve_lower(L, f - mean_total(X))
+        mu = self.mean_func(Xnew) + tt.dot(tt.transpose(A), v)
+        Kss = self.cov_func(Xnew)
+        cov = Kss - tt.dot(tt.transpose(A), A)
+        return mu, cov
+
+    def conditional(self, name, Xnew, given={}, **kwargs):
+        R"""
+	Returns the conditional distribution evaluated over new input
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py:166]: Indent

File: pymc3/gp/gp.py [hunk 14]
```diff
@@ -1,151 +1,753 @@
...
+        covT = (self.nu + beta - 2)/(nu2 - 2) * cov
+        return nu2, mu, covT
 
-        if X is not None:
-            self.X = X
-            self.mean = self.mode = self.M(X)
-            kwargs.setdefault("shape", X.squeeze().shape)
+    def conditional(self, name, Xnew, **kwargs):
+        R"""
+	Returns the conditional distribution evaluated over new input
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py:286]: Indent

File: pymc3/gp/gp.py [hunk 15]
```diff
@@ -1,151 +1,753 @@
...
+        c = solve_lower(L_B, tt.dot(A, r_l))
+        constant = 0.5 * X.shape[0] * tt.log(2.0 * np.pi)
+        logdet = 0.5 * tt.sum(tt.log(Lamd)) + tt.sum(tt.log(tt.diag(L_B)))
+        quadratic = 0.5 * (tt.dot(r, r_l) - tt.dot(c, c))
+        return -1.0 * (constant + logdet + quadratic + trace)
 
-    data = [v for v in model.observed_RVs if v.name==gp.name][0].data
+    def marginal_likelihood(self, name, X, Xu, y, sigma, is_observed=True, **kwargs):
+        R"""
+	Returns the approximate marginal likelihood distribution, given the input
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py:641]: Indent

**Comment by fonnesbeck** [comment]
Looks great. Only minor cosmetic comments from me. Go ahead and merge.

**Comment by fonnesbeck** [comment]
Since the issues I cited are just docstrings, 'm going to merge this, and just submit a PR to address them.

**Comment by fonnesbeck** [comment]
Thanks, @bwengals for the massive contribution, and congratulations on a very successful GSoC!

**Comment by bwengals** [comment]
Wow merged, alright!  Thank everyone for the help.  Im at the coast, but will back to clean these issues up in 2 days.

**Comment by twiecki** [comment]
This is an incredible amount of great work, congratulations!

**Comment by ferrine** [comment]
Congrats!

**Comment by junpenglao** [comment]
Great work @bwengals!!!

**Comment by AustinRochford** [comment]
innovate_computer

**Comment by ericmjl** [comment]
Woohoo!!

**Comment by usptact** [comment]
Amazing work @bwengals !
